contracts/escrow-fee-distributor/src/contract.rs

```diff
@@ -0,0 +1,881 @@
+        }
+
+        week_cursor += WEEK;
+    }
+
+    TIME_CURSOR_OF.save(deps.storage, addr.clone(), &week_cursor)?;
+
+    CLAIMED.save(
+        deps.storage,
+        &vec![Claimed {
+            recipient: addr,
+            amount: to_distribute,
+            claim_period: get_period(week_cursor),
+            max_period: user_lock_info.end,
+        }],
+    )?;
+
+    Ok(to_distribute)
+}
+
+/// ## Description
+/// Updates general settings. Returns an [`ContractError`] on failure or the following [`Config`]
+/// data will be updated if successful.
+fn update_config(
+    deps: DepsMut,
+    info: MessageInfo,
+    max_limit_accounts_of_claim: Option<u64>,
+    can_checkpoint_token: Option<bool>,
+) -> Result<Response, ContractError> {
+    let mut attributes = vec![attr("action", "set_config")];
+    let mut config: Config = CONFIG.load(deps.storage)?;
+
+    if config.is_killed {
+        return Err(ContractError::ContractIsKilled {});
+    }
+
+    if info.sender != config.owner {
+        return Err(ContractError::Unauthorized {});
+    }
+
+    if let Some(can_checkpoint_token) = can_checkpoint_token {
+        config.can_checkpoint_token = can_checkpoint_token;
+        attributes.push(Attribute::new(
+            "can_checkpoint_token",
+            can_checkpoint_token.to_string(),
+        ));
+    };
+
+    if let Some(max_limit_accounts_of_claim) = max_limit_accounts_of_claim {
+        config.max_limit_accounts_of_claim = max_limit_accounts_of_claim;
+        attributes.push(Attribute::new(
+            "max_limit_accounts_of_claim",
+            max_limit_accounts_of_claim.to_string(),
+        ));
+    };
+
+    CONFIG.save(deps.storage, &config)?;
+
+    Ok(Response::new().add_attributes(attributes))
+}
+
+/// ## Description
+/// Available the query messages of the contract.
+/// ## Queries
+/// * **QueryMsg::Config {}** Returns the base controls configs that contains in the [`Config`] object.
+///
+/// * **QueryMsg::AstroRecipientsPerWeek {}** Returns the list of accounts who will get ASTRO fees
+/// every week in the [`RecipientsPerWeekResponse`] object.
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn query(deps: Deps, env: Env, msg: QueryMsg) -> StdResult<Binary> {
+    match msg {
+        QueryMsg::FetchUserBalanceByTimestamp { user, timestamp } => {
+            Ok(to_binary(&query_user_balance(deps, env, user, timestamp)?)?)
+        }
+        QueryMsg::Config {} => Ok(to_binary(&query_config(deps)?)?),
+        QueryMsg::AstroRecipientsPerWeek {} => Ok(to_binary(&query_recipients_per_weeks(deps)?)?),
+        QueryMsg::VotingSupplyPerWeek {} => Ok(to_binary(&query_voting_supply_per_week(deps)?)?),
+        QueryMsg::FeeTokensPerWeek {} => Ok(to_binary(&query_tokens_per_week(deps)?)?),
+    }
+}
+
+fn query_voting_supply_per_week(deps: Deps) -> StdResult<Vec<Uint128>> {
+    let mut result: Vec<Uint128> = vec![];
+    for x in VOTING_SUPPLY_PER_WEEK.keys(deps.storage, None, None, Order::Ascending) {
```

Comment:
Maybe we need to add pagination here? For example, add parameter `weeks: Option<u64>` and show last 10 weeks only if None.